src/login/connect_server.h

```diff
@@ -0,0 +1,2257 @@
+    std::vector<uint32> sandoriaStartingZones = { 0xE6, 0xE7, 0xE8 };
+    std::vector<uint32> windurstStartingZones = { 0xEE, 0xF0, 0xF1 };
+
+    switch (createchar.m_nation)
+    {
+        case 0x02: // windy start
+            createchar.m_zone = windurstStartingZones[rand() % 3];
+            break;
+        case 0x01: // bastok start
+            createchar.m_zone = bastokStartingZones[rand() % 3];
+            break;
+        case 0x00: // sandy start
+            createchar.m_zone = sandoriaStartingZones[rand() % 3];
+            break;
+    }
+
+    const char* fmtQuery = "SELECT max(charid) FROM chars";
+
+    if (sql->Query(fmtQuery) == SQL_ERROR)
+    {
+        return -1;
+    }
+
+    uint32 CharID = 0;
+
+    if (sql->NumRows() != 0)
+    {
+        sql->NextRow();
+
+        CharID = sql->GetUIntData(0) + 1;
+    }
+
+    if (saveCharacter(session.accountID, CharID, &createchar) == -1)
+    {
+        return -1;
+    }
+
+    ShowDebug(fmt::format("char<{}> successfully saved", createchar.m_name));
+    return 0;
+};
+
+// Interaction with xiloader, port 54231
+class auth_session : public handler_session
+{
+public:
+    auth_session(asio::ssl::stream<asio::ip::tcp::socket> socket)
+    : handler_session(std::move(socket))
+    {
+        DebugSockets("auth_session");
+    }
+
+public:
+    void start()
+    {
+        auto self(shared_from_this());
+        // clang-format off
+
+        socket_.async_handshake(asio::ssl::stream_base::server,
+        [this, self](std::error_code ec)
+        {
+            if (!ec)
+            {
+                do_read();
+            }
+            else
+            {
+                ShowWarning(fmt::format("Error: ({}), {}", ec.value(), ec.message()));
+                ShowWarning("Failed to handshake!");
+                socket_.next_layer().close();
+            }
+        });
+        // clang-format on
+    }
+
+protected:
+    void do_read()
+    {
+        auto self(shared_from_this());
+        // clang-format off
+        socket_.async_read_some(asio::buffer(data_, max_length),
+        [this, self](std::error_code ec, std::size_t length)
+        {
+            if (!ec)
+            {
+                read_func();
+            }
+            else
+            {
+                DebugSockets(ec.message());
+                handle_error(ec, self);
+            }
+        });
+        // clang-format on
+    }
+
+    void read_func() override
+    {
+        auto newModeFlag = ref<uint8>(data_, 0) == 0xFF;
+        if (!newModeFlag)
+        {
+            ShowDebug("Old xiloader connected. Not supported.");
+            ref<uint8>(data_, 0) = LOGIN_ERROR;
+            do_write(1);
+            return;
+        }
+
+        /*auto componentFlags = ref<uint64>(data_, 1);
+
+        if (componentFlags & )
+        {
+            // ...
+        }*/
+
+        char usernameBuffer[17] = {};
+        char passwordBuffer[17] = {};
```

Comment:
While we're doing a breaking change, should/could we increase the max password length to 32? and how hard/clearly do we enforce it here and on xiloader?

These lengths are used a few times, extract them out somewhere?

Reply:
password is 32 characters now